Language Server holds ref assembly DLLs open, blocking MSBuild CopyRefAssembly

## Summary

The G# Language Server (running via the VS Code extension) keeps ref assembly DLLs open (memory-mapped + file descriptor), which causes `dotnet build` to fail with `IOException: The process cannot access the file ... because it is being used by another process`.

## Reproduction

1. Open a G# workspace in VS Code (e.g. the Oahu solution).
2. Build the solution once so ref assemblies exist in `obj/Debug/net10.0/ref/`.
3. From a terminal, run `dotnet build Oahu.sln`.
4. MSBuild fails at the `CopyRefAssembly` task for projects whose ref DLLs the language server has open.

## Observed behavior

```
error MSB3883: Unexpected exception:
IOException: The process cannot access the file
  '.../obj/Debug/net10.0/ref/Oahu.Foundation.dll'
  because it is being used by another process.
   at Microsoft.CodeAnalysis.BuildTasks.CopyRefAssembly.Copy()
```

`lsof` confirms the language server process holds the files:

```
COMMAND   PID  USER        FD   TYPE  NAME
dotnet  62551 davidobando  txt  REG   .../obj/Debug/net10.0/ref/Oahu.Foundation.dll
dotnet  62551 davidobando  452r REG   .../obj/Debug/net10.0/ref/Oahu.Foundation.dll
dotnet  62551 davidobando  txt  REG   .../obj/Debug/net10.0/ref/Oahu.Decrypt.dll
dotnet  62551 davidobando  450r REG   .../obj/Debug/net10.0/ref/Oahu.Decrypt.dll
```

PID 62551 is:
```
dotnet .../.vscode/extensions/gsharplang.vscode-gsharp-0.2.13/.server/GSharp.LanguageServer.dll --stdio
```

## Expected behavior

The language server should either:
- Open ref assemblies with `FileShare.ReadWrite | FileShare.Delete` so MSBuild can overwrite them, or
- Use shadow-copy / memory-only loading (e.g. `MetadataReference.CreateFromStream`) so the files are never locked on disk.

## Environment

- macOS (Darwin)
- .NET SDK 10.0.301
- G# VS Code extension v0.2.13
- G# Language Server from Gsharp.NET.Sdk 0.2.1

## Workaround

Terminate the language server process before building, or close the VS Code window.

--- a/src/Core/CodeAnalysis/Symbols/ReferenceResolver.cs
+++ b/src/Core/CodeAnalysis/Symbols/ReferenceResolver.cs
@@ -206,7 +206,7 @@ public sealed class ReferenceResolver : IDisposable
         {
             try
             {
-                var asm = mlc.LoadFromAssemblyPath(path);
+                var asm = resolver.LoadFromPath(mlc, path);
                 if (asm != null && seen.Add(asm.GetName().FullName))
                 {
                     builder.Add(asm);
@@ -247,7 +247,7 @@ public sealed class ReferenceResolver : IDisposable
 
                 try
                 {
-                    var asm = mlc.LoadFromAssemblyPath(host);
+                    var asm = resolver.LoadFromPath(mlc, host);
                     if (asm != null && seen.Add(asm.GetName().FullName))
                     {
                         builder.Add(asm);
@@ -737,13 +737,33 @@ public sealed class ReferenceResolver : IDisposable
     /// </summary>
     private sealed class FallbackMetadataAssemblyResolver : MetadataAssemblyResolver
     {
-        private readonly PathAssemblyResolver inner;
+        private readonly Dictionary<string, byte[]> assemblyBytes = new(StringComparer.OrdinalIgnoreCase);
         private readonly HashSet<string> unresolved = new(StringComparer.OrdinalIgnoreCase);
         private readonly object gate = new();
 
         public FallbackMetadataAssemblyResolver(IEnumerable<string> paths)
         {
-            inner = new PathAssemblyResolver(paths);
+            // Read each assembly into memory so no file handles are held open.
+            // This prevents MSBuild CopyRefAssembly from being blocked (#853).
+            foreach (var path in paths)
+            {
+                var fileName = Path.GetFileNameWithoutExtension(path);
+                if (!assemblyBytes.ContainsKey(fileName))
+                {
+                    try
+                    {
+                        using var fs = new FileStream(path, FileMode.Open, FileAccess.Read, FileShare.ReadWrite | FileShare.Delete);
+                        var bytes = new byte[fs.Length];
+                        fs.ReadExactly(bytes);
+                        assemblyBytes[fileName] = bytes;
+                    }
+                    catch (Exception ex) when (
+                        ex is IOException or UnauthorizedAccessException or BadImageFormatException)
+                    {
+                        // Skip files that can't be read.
+                    }
+                }
+            }
         }
 
         public IEnumerable<string> UnresolvedSimpleNames
@@ -757,30 +777,48 @@ public sealed class ReferenceResolver : IDisposable
             }
         }
 
-        public override Assembly Resolve(MetadataLoadContext context, AssemblyName assemblyName)
+        /// <summary>
+        /// Loads an assembly by path into the given <see cref="MetadataLoadContext"/>
+        /// using in-memory bytes (no file handle held open).
+        /// </summary>
+        public Assembly LoadFromPath(MetadataLoadContext context, string path)
         {
-            try
+            var fileName = Path.GetFileNameWithoutExtension(path);
+            if (assemblyBytes.TryGetValue(fileName, out var bytes))
             {
-                var resolved = inner.Resolve(context, assemblyName);
-                if (resolved != null)
-                {
-                    return resolved;
-                }
-            }
-            catch (Exception ex) when (
-                ex is FileNotFoundException
-                    or FileLoadException
-                    or BadImageFormatException)
-            {
-                // Fall through to record-and-degrade below.
+                return context.LoadFromByteArray(bytes);
             }
 
-            var name = assemblyName?.Name;
-            if (!string.IsNullOrEmpty(name))
+            // Fallback: read from disk with sharing flags.
+            using var fs = new FileStream(path, FileMode.Open, FileAccess.Read, FileShare.ReadWrite | FileShare.Delete);
+            var buffer = new byte[fs.Length];
+            fs.ReadExactly(buffer);
+            return context.LoadFromByteArray(buffer);
+        }
+
+        public override Assembly Resolve(MetadataLoadContext context, AssemblyName assemblyName)
+        {
+            var simpleName = assemblyName?.Name;
+            if (!string.IsNullOrEmpty(simpleName) && assemblyBytes.TryGetValue(simpleName, out var bytes))
+            {
+                try
+                {
+                    return context.LoadFromByteArray(bytes);
+                }
+                catch (Exception ex) when (
+                    ex is FileNotFoundException
+                        or FileLoadException
+                        or BadImageFormatException)
+                {
+                    // Fall through to record-and-degrade below.
+                }
+            }
+
+            if (!string.IsNullOrEmpty(simpleName))
             {
                 lock (gate)
                 {
-                    unresolved.Add(name);
+                    unresolved.Add(simpleName);
                 }
             }
 

--- a/src/LanguageServer/PdbSourceLocator.cs
+++ b/src/LanguageServer/PdbSourceLocator.cs
@@ -171,7 +171,7 @@ public static class PdbSourceLocator
 
         try
         {
-            using var stream = new FileStream(assemblyFilePath, FileMode.Open, FileAccess.Read, FileShare.Read);
+            using var stream = new FileStream(assemblyFilePath, FileMode.Open, FileAccess.Read, FileShare.ReadWrite | FileShare.Delete);
             using var peReader = new PEReader(stream);
             if (!peReader.HasMetadata)
             {
@@ -300,7 +300,7 @@ public static class PdbSourceLocator
             }
 
             // Probe the PE for an embedded PDB.
-            using var stream = new FileStream(assemblyFilePath, FileMode.Open, FileAccess.Read, FileShare.Read);
+            using var stream = new FileStream(assemblyFilePath, FileMode.Open, FileAccess.Read, FileShare.ReadWrite | FileShare.Delete);
             using var peReader = new PEReader(stream);
             if (!peReader.HasMetadata)
             {

